### PR TITLE
Check universes are declared

### PR DESCRIPTION
**Pull request by mattam82**
This is on top of PR #890  [merged]
This is a tentative to ensure during typechecking in the kernel that universes in the terms are properly declared, even though they don't necessarily appear in constraints. Currently the kernel accepts term with undeclared universes and fails at their first use. I fear this is going to be costly (esp. on universe polymorphic code) and we should maybe do this more globally as a well-formedness test before typing, collecting all the universes of the term and checking their well-formedness at that time (maybe also checking that there is no algebraic universe in the wrong place at that time).

### CODE REVIEW

**Comment by ppedrot** [comment]
Instead of doing this check at typing time, I think it should go globally as one separate pass performed in the `infer` function. This is the one function that delimits unsafe from type-checked terms, so it would make sense to put it there.

**Comment by mattam82** [comment]
Alright, I will try that version.

**Comment by mattam82** [comment]
Apparently it does catch bugs on travis :)

**Comment by ppedrot** [comment]
Arg...

**Comment by mattam82** [comment]
I think it's better to catch those than not though

File: parsing/g_constr.ml4
Comment (SkySkimmer) [inline parsing/g_constr.ml4]: It's kind of strange to have anonymous not mean underscore for universes.

File: kernel/constr.ml
Comment (SkySkimmer) [inline kernel/constr.ml]: Isn't this function the same as Univops.universes_of_constr? Except this uses LSet.union where the other uses a fold over the instance.

Comment (mattam82) [reply kernel/constr.ml]: Right, that's duplicated

File: parsing/g_constr.ml4
Comment (mattam82) [reply parsing/g_constr.ml4]: I'll revert the naming

**Comment by mattam82** [comment]
Fixed the issues raised by @SkySkimmer in PR#890 and here

**Comment by SkySkimmer** [comment]
This may also be a good place to check that algebraic universes don't end up in terms.

**Comment by SkySkimmer** [comment]
It should be relatively simple to rebase (ie just cherry pick the commit) and merge this PR now that #890 is in.

**Comment by maximedenes** [comment]
Does the checker implement such a check already? If not, I believe it should.

**Comment by SkySkimmer** [comment]
@mattam82 would you mind if I try to rebase & fix this PR and push to your branch?

**Comment by mattam82** [comment]
@SkySkimmer I wouldn't mind :)

**Comment by SkySkimmer** [comment]
Rebased.
Travis showed some errors before that so let's see.

**Comment by SkySkimmer** [comment]
There's some error possibly related to Program or automatic scheme declaration:
```coq
Require Import Coq.Classes.Morphisms Coq.Program.Program.
Set Universe Polymorphism.

Module with_vars.
(* Some random utilities: *)

Lemma sum_assoc {A B C}: (A+B)+C -> A+(B+C). intuition. Defined.
Lemma bla {A B C}: (A+B) -> A+(B+C). intuition. Defined.

(* Again our example term language, this time without plus/one (they're boring), but with Var
 added: *)

Inductive Expr (V: Type) := Mult (a b: Expr V).

Implicit Arguments Mult [[V]].


Definition Value := nat.
Definition Vars V := V -> Value.

Fixpoint eval {V} (vs: Vars V) (e: Expr V): Value :=
  match e with
  | Mult a b => eval vs a * eval vs b
  end.

Instance eval_proper V: Proper (pointwise_relation _ eq ==> eq ==> eq) (@eval V).
Proof. Admitted.

(* Some simple combinators for variable packs: *)

Definition merge {A B} (a: Vars A) (b: Vars B): Vars (A+B) :=
  fun i => match i with inl j => a j | inr j => b j end.

Definition map_var {V W: Type} (f: V -> W): Expr V -> Expr W. Admitted.

Lemma eval_map_var {V W} (f: V -> W) v e:
  eval v (map_var f e) = eval (v ∘ f) e.
Proof. Admitted.


Section Quote.

  Class Quote {V} (l: Vars V) (n: Value) {V'} (r: Vars V') :=
    { quote: Expr (V + V')
    ; eval_quote: @eval (V+V') (merge l r) quote = n }.

  Implicit Arguments quote [[V] [l] [V'] [r] [Quote]].

  Global Program Instance quote_mult V (v: Vars V) n V' (v': Vars V') m V'' (v'': Vars V'')
    `{!Quote v n v'} `{!Quote (merge v v') m v''}: Quote v (n * m) (merge v' v'') :=
      { quote := Mult (map_var bla (quote n)) (map_var sum_assoc (quote m)) }.

  Next Obligation. Proof with auto.
   destruct Quote0, Quote1.
   subst. simpl.
   do 2 rewrite eval_map_var.
   f_equal; apply eval_proper; auto; intro; intuition.
  Qed. (* this PR: undeclared universe *)

  Set Printing Universes.
  Print quote_mult_obligation_1.
  (* with master:
quote_mult_obligation_1@{Top.88 Top.89 Top.91 Top.97 Top.99 Top.106 Top.107 Top.110 Top.112
Top.120} = 
let Top_with_vars_Quote_internal_eq_rew_dep :=
  fun (A : Type@{Coq.Init.Logic.8}) (x : A) (P : forall a : A, x = a -> Type@{Top.122} (* <- XXX THIS ONE *) )
    (f : P x eq_refl) (y : A) (e : x = y) =>
  match e as e0 in (_ = y0) return (P y0 e0) with
  | eq_refl => f
  end in
... *)
```
This is from mathclasses (I turned universe polymorphism on for the example because that way we easily see that it's undeclared, but the error still occurs without polymorphism).

There's another error in equations, also in an obligation but beyond that I don't know if it's related.
```
File "examples/polynomials.v", line 370, characters 2-10:
Error: Undeclared universe: polynomials.2242 (maybe a bugged tactic).
```

**Comment by Zimmi48** [comment]
@SkySkimmer Did you intend to take over this PR?

**Comment by SkySkimmer** [comment]
I tried but it didn't work out (I am not good at Program debugging)

**Comment by SkySkimmer** [comment]
Also shouldn't the checker be running this check too?

**Comment by Zimmi48** [comment]
Maybe having only the checker do this check would be a good, simpler, first step...

**Comment by SkySkimmer** [comment]
Reduced example
```coq
Require Import Coq.Program.Program.
Set Universe Polymorphism.
Set Printing Universes.

Axiom ALL : forall {T:Prop}, T.

Inductive Expr : Set := E (a : Expr).

Parameter Value : Set.

Fixpoint eval (e: Expr): Value :=
  match e with
  | E a => eval a
  end.

Class Quote (n: Value) : Set :=
  { quote: Expr
    ; eval_quote: eval quote = n }.

Program Definition quote_mult n
        `{!Quote n} : Quote n :=
  {| quote := E (quote (n:=n)) |}.


Next Obligation.
Proof.
  destruct Quote0 as [q eq].
  subst.
  apply ALL.
Qed.
Print quote_mult_obligation_1.
(* quote_mult_obligation_1@{} =
let Top_internal_eq_rew_dep :=
  fun (A : Type@{Coq.Init.Logic.8}) (x : A) (P : forall a : A, x = a -> Type@{Top.5} (* <- XXX *))
    (f : P x eq_refl) (y : A) (e : x = y) =>
  match e as e0 in (_ = y0) return (P y0 e0) with
  | eq_refl => f
  end in
fun (n : Value) (Quote0 : Quote n) =>
match Quote0 as q return (eval quote = n) with
| {| quote := q; eval_quote := eq0 |} =>
    Top_internal_eq_rew_dep Value (eval q) (fun (n0 : Value) (eq1 : eval q = n0) => eval quote = n0)
      ALL n eq0
end
     : forall (n : Value) (Quote0 : Quote n), eval (E quote) = n

quote_mult_obligation_1 is universe polymorphic
*)
```

File: kernel/typeops.ml
Comment (ppedrot) [inline kernel/typeops.ml]: I know this is just moving code, but it's fishy to add the global universes to the one appearing in the term. They're not quite part of the term, sometimes they're even half garbage (e.g. side-effects). It's fine because it's an overapproximation, but still.

**Comment by mattam82** [comment]
You’re right. I think this was added to handle the case of side-effects
precisely, maybe it should be investigated some more. I would bet something
breaks if we remove that line.
— Matthieu

**Comment by SkySkimmer** [comment]
> I would bet something breaks if we remove that line.

I tried and it doesn't break (any more?).
It seems in https://github.com/coq/coq/pull/6324#issuecomment-349621501 
>New approach which I feel is more principled: universes_of_constr includes the universes of monomorphic constants appearing in the term.

I actually kept the old approach alongside? I don't remember.

**Review by maximedenes — DISMISSED** [review]
I request the PR to be fixed. Ping me again once it is done. I'm making this review to improve my stats as collected by @ejgallego :)

**Comment by SkySkimmer** [comment]
Can you add `fix #7900` to the commit message?

**Comment by SkySkimmer** [comment]
Also the test suite file should be 7900.v not 891.v.

**Comment by ppedrot** [comment]
@SkySkimmer You can edit the message yourself (at least I can).

**Comment by SkySkimmer** [comment]
And the test suite is broken.

**Comment by mattam82** [comment]
Indeed it was more subtle than I thought. Should be better now, and the merge_context_set options are  commented.

**Comment by SkySkimmer** [comment]
Looks like this was ready and waiting for review approval/merging so I reset the requests for review.

**Review by SkySkimmer — APPROVED** [review]

**Comment by maximedenes** [comment]
@SkySkimmer can you be the assignee for this one? I'm not competent, I'm afraid.

**Comment by SkySkimmer** [comment]
Sure I can take it.
>I'm not competent, I'm afraid.

Do you mean about the part outside the kernel fixing #7900? The kernel part looks straightforward to me.
TBH I don't especially understand this side effect stuff either. I don't know who does but @ppedrot is marked as codeowner so maybe he'll comment.

**Review by ppedrot — APPROVED** [review]
A must-have in 8.9, that will prevent a lot of fancy and untestable soundness bugs.

**Comment by SkySkimmer** [comment]
Indeed undeclared universes could re-emerge the soundness bug of #8341 for instance.

**Pull request merged**
